Problem with variadic macros

When compiling Libint 2.4.1 using GCC 6.3.0, on one Linux system I get the error message

 libint/build_libint.cc:38:4: error: #error "your compiler does not provide variadic macros (but does support C++11), something is seriously broken, please create an issue at https://github.com/evaleev/libint/issues"

whereas on another Linux computer the code builds fine with the same version of GCC.

--- a/CHANGES
+++ b/CHANGES
@@ -9,7 +9,7 @@ Following is a brief summary of changes made in each release of Libint.
   - resolved issue 95
   - minor fixes for C++17 (includes PR 96 HT https://github.com/mfherbst)
   - [generated lib only] ensure system boost/preprocessor.hpp is more recent than 1.57; reverts to bundled boost is system boost is too old
-  - resolved issue 97: ensures that libint compiler has boost is 1.57 or older
+  - resolved issue 97: checks variadic macro support in boost.preprocessor for the libint compiler
 
 - 2017-11-21: 2.4.2
   - fixed vector.h for use on non-x86 platforms

--- a/CHANGES
+++ b/CHANGES
@@ -3,12 +3,13 @@
 
 Following is a brief summary of changes made in each release of Libint.
 
-- 2018-xx-yy: 2.4.3
+- 2018-xx-yy: 2.5.0
   - resolved issue 90
   - resolved issue 91: fixed formatting errors in def2-sv(p)-c, def2-tzvp-c, and def2-tzvpp-c basis set definitions
   - resolved issue 95
   - minor fixes for C++17 (includes PR 96 HT https://github.com/mfherbst)
-  - ensures system boost/preprocessor.hpp is more recent than 1.57; reverts to bundled boost is system boost is too old
+  - [generated lib only] ensure system boost/preprocessor.hpp is more recent than 1.57; reverts to bundled boost is system boost is too old
+  - resolved issue 97: ensures that libint compiler has boost is 1.57 or older
 
 - 2017-11-21: 2.4.2
   - fixed vector.h for use on non-x86 platforms

--- a/configure.ac
+++ b/configure.ac
@@ -1307,8 +1307,28 @@ AC_ARG_ENABLE(mpfr,
 
 dnl ------------------ Check for Boost library -----------------------
 
-dnl Boost.Preprocessor requires 1.57
-AX_BOOST_BASE([1.57],, [AC_MSG_ERROR([libint compiler needs Boost 1.57 or later, download from https://www.boost.org/users/download/ , unpack (no need to install), and add -I/path/to/boost to CPPFLAGS])])
+dnl Boost.Preprocessor requires 1.30 (which introduced MPL; Preprocessor was introduced in 1.29)
+AX_BOOST_BASE([1.30],, [AC_MSG_ERROR([libint compiler needs Boost 1.30 or later, download from https://www.boost.org/users/download/ , unpack (no need to install), and add -I/path/to/boost to CPPFLAGS])])
+
+AC_MSG_CHECKING([for variadic macro support in Boost.Preprocessor])
+AC_COMPILE_IFELSE(
+  [AC_LANG_PROGRAM(
+    [[#include <boost/preprocessor.hpp>
+      #if not BOOST_PP_VARIADICS  // no variadic macros? your compiler is out of date! (should not be possible since variadic macros are part of C++11)
+      #  error "your compiler does not provide variadic macros (but does support C++11), something is seriously broken, please create an issue at https://github.com/evaleev/libint/issues"
+      #endif
+    ]],
+    [[
+    ]]
+   )
+  ],
+  [AC_MSG_RESULT([yes])
+  ],
+  [
+    AC_MSG_RESULT([no])
+    AC_MSG_ERROR([Boost.Preprocessor with variadic macro support is required. Download latest boost from https://www.boost.org/users/download/ , unpack (no need to install), and add -I/path/to/boost to CPPFLAGS])
+  ]
+)
 
 AC_MSG_CHECKING([for Boost::mpl::for_each])
 AC_COMPILE_IFELSE(
@@ -1330,7 +1350,7 @@ AC_COMPILE_IFELSE(
   ],
   [
     AC_MSG_RESULT([no])
-    AC_MSG_ERROR([Boost::mpl is required. Please use the latest Boost library from boost.org and/or most standard-compliant C++ compiler.])
+    AC_MSG_ERROR([Boost::mpl is required. Download latest boost from https://www.boost.org/users/download/ , unpack (no need to install), and add -I/path/to/boost to CPPFLAGS])
   ]
 )
 

--- a/configure.ac
+++ b/configure.ac
@@ -1305,50 +1305,10 @@ AC_ARG_ENABLE(mpfr,
  ]
 )
 
-dnl ------------------ Check for shared_ptr -----------------------
+dnl ------------------ Check for Boost library -----------------------
 
-HAVE_SHARED_PTR=no
-AC_MSG_CHECKING([for Boost::shared_ptr])
-AC_COMPILE_IFELSE(
-  [AC_LANG_PROGRAM(
-    [[#include <boost/shared_ptr.hpp>
-    ]],
-    [[
-     boost::shared_ptr<int> a(new int);
-     boost::shared_ptr<int> b = a;
-    ]]
-   )
-  ],
-  [AC_MSG_RESULT([yes])
-   HAVE_SHARED_PTR=boost
-   AC_DEFINE(HAVE_SHARED_PTR_IN_BOOST)
-  ],
-  [
-    AC_MSG_RESULT([no])
-  ]
-)
-
-if test "X$HAVE_SHARED_PTR" = "Xno"; then
-  AC_MSG_CHECKING([for std::shared_ptr])
-  AC_COMPILE_IFELSE(
-    [AC_LANG_PROGRAM(
-      [[#include <memory>
-      ]],
-      [[
-        std::shared_ptr<int> a(new int);
-        std::shared_ptr<int> b = a;
-      ]]
-     )
-    ],
-    [AC_MSG_RESULT([yes])
-     HAVE_SHARED_PTR=std
-    ],
-    [
-      AC_MSG_RESULT([no])
-      AC_MSG_ERROR([shared_ptr is required. Please download Boost from boost.org or use C++ compiler that supports shared_ptr.])
-    ]
-  )
-fi
+dnl Boost.Preprocessor requires 1.57
+AX_BOOST_BASE([1.57],, [AC_MSG_ERROR([libint compiler needs Boost 1.57 or later, download from https://www.boost.org/users/download/ , unpack (no need to install), and add -I/path/to/boost to CPPFLAGS])])
 
 AC_MSG_CHECKING([for Boost::mpl::for_each])
 AC_COMPILE_IFELSE(
@@ -1374,6 +1334,51 @@ AC_COMPILE_IFELSE(
   ]
 )
 
+dnl ------------------ Check for shared_ptr -----------------------
+
+HAVE_SHARED_PTR=no
+AC_MSG_CHECKING([for std::shared_ptr])
+AC_COMPILE_IFELSE(
+  [AC_LANG_PROGRAM(
+      [[#include <memory>
+      ]],
+      [[
+        std::shared_ptr<int> a(new int);
+        std::shared_ptr<int> b = a;
+      ]]
+     )
+    ],
+    [AC_MSG_RESULT([yes])
+     HAVE_SHARED_PTR=std
+    ],
+    [
+      AC_MSG_RESULT([no])
+    ]
+  )
+
+if test "X$HAVE_SHARED_PTR" = "Xno"; then
+  AC_MSG_CHECKING([for Boost::shared_ptr])
+  AC_COMPILE_IFELSE(
+    [AC_LANG_PROGRAM(
+      [[#include <boost/shared_ptr.hpp>
+      ]],
+      [[
+       boost::shared_ptr<int> a(new int);
+       boost::shared_ptr<int> b = a;
+      ]]
+     )
+    ],
+    [AC_MSG_RESULT([yes])
+     HAVE_SHARED_PTR=boost
+     AC_DEFINE(HAVE_SHARED_PTR_IN_BOOST)
+    ],
+    [
+      AC_MSG_RESULT([no])
+      AC_MSG_ERROR([shared_ptr is required. Please download Boost from boost.org or use C++ compiler that supports std::shared_ptr.])
+    ]
+  )
+fi
+
 dnl ------------------ Check for Eigen library -----------------------
 
 ACX_CHECK_EIGEN
@@ -1383,7 +1388,7 @@ fi
 
 dnl -- Checks for typedefs, structures, and compiler characteristics. --
 AC_TYPE_SIZE_T
-
+``
 CPPFLAGS="$CPPFLAGS" CXX="$CXX" CXXFLAGS="$CXXFLAGS" \
 LD="$LD" LDFLAGS="$LDFLAGS" LIBS="-lm $LIBS $FLIBS" \
 LN_S="$LN_S" NM="$NM" RANLIB="$RANLIB"

--- a/lib/autoconf/ax_boost_base.m4
+++ b/lib/autoconf/ax_boost_base.m4
@@ -1,0 +1,301 @@
+# ===========================================================================
+#      https://www.gnu.org/software/autoconf-archive/ax_boost_base.html
+# ===========================================================================
+#
+# SYNOPSIS
+#
+#   AX_BOOST_BASE([MINIMUM-VERSION], [ACTION-IF-FOUND], [ACTION-IF-NOT-FOUND])
+#
+# DESCRIPTION
+#
+#   Test for the Boost C++ libraries of a particular version (or newer)
+#
+#   If no path to the installed boost library is given the macro searchs
+#   under /usr, /usr/local, /opt and /opt/local and evaluates the
+#   $BOOST_ROOT environment variable. Further documentation is available at
+#   <http://randspringer.de/boost/index.html>.
+#
+#   This macro calls:
+#
+#     AC_SUBST(BOOST_CPPFLAGS) / AC_SUBST(BOOST_LDFLAGS)
+#
+#   And sets:
+#
+#     HAVE_BOOST
+#
+# LICENSE
+#
+#   Copyright (c) 2008 Thomas Porschberg <thomas@randspringer.de>
+#   Copyright (c) 2009 Peter Adolphs
+#
+#   Copying and distribution of this file, with or without modification, are
+#   permitted in any medium without royalty provided the copyright notice
+#   and this notice are preserved. This file is offered as-is, without any
+#   warranty.
+
+#serial 44
+
+# example boost program (need to pass version)
+m4_define([_AX_BOOST_BASE_PROGRAM],
+          [AC_LANG_PROGRAM([[
+#include <boost/version.hpp>
+]],[[
+(void) ((void)sizeof(char[1 - 2*!!((BOOST_VERSION) < ($1))]));
+]])])
+
+AC_DEFUN([AX_BOOST_BASE],
+[
+AC_ARG_WITH([boost],
+  [AS_HELP_STRING([--with-boost@<:@=ARG@:>@],
+    [use Boost library from a standard location (ARG=yes),
+     from the specified location (ARG=<path>),
+     or disable it (ARG=no)
+     @<:@ARG=yes@:>@ ])],
+    [
+     AS_CASE([$withval],
+       [no],[want_boost="no";_AX_BOOST_BASE_boost_path=""],
+       [yes],[want_boost="yes";_AX_BOOST_BASE_boost_path=""],
+       [want_boost="yes";_AX_BOOST_BASE_boost_path="$withval"])
+    ],
+    [want_boost="yes"])
+
+
+AC_ARG_WITH([boost-libdir],
+  [AS_HELP_STRING([--with-boost-libdir=LIB_DIR],
+    [Force given directory for boost libraries.
+     Note that this will override library path detection,
+     so use this parameter only if default library detection fails
+     and you know exactly where your boost libraries are located.])],
+  [
+   AS_IF([test -d "$withval"],
+         [_AX_BOOST_BASE_boost_lib_path="$withval"],
+    [AC_MSG_ERROR([--with-boost-libdir expected directory name])])
+  ],
+  [_AX_BOOST_BASE_boost_lib_path=""])
+
+BOOST_LDFLAGS=""
+BOOST_CPPFLAGS=""
+AS_IF([test "x$want_boost" = "xyes"],
+      [_AX_BOOST_BASE_RUNDETECT([$1],[$2],[$3])])
+AC_SUBST(BOOST_CPPFLAGS)
+AC_SUBST(BOOST_LDFLAGS)
+])
+
+
+# convert a version string in $2 to numeric and affect to polymorphic var $1
+AC_DEFUN([_AX_BOOST_BASE_TONUMERICVERSION],[
+  AS_IF([test "x$2" = "x"],[_AX_BOOST_BASE_TONUMERICVERSION_req="1.20.0"],[_AX_BOOST_BASE_TONUMERICVERSION_req="$2"])
+  _AX_BOOST_BASE_TONUMERICVERSION_req_shorten=`expr $_AX_BOOST_BASE_TONUMERICVERSION_req : '\([[0-9]]*\.[[0-9]]*\)'`
+  _AX_BOOST_BASE_TONUMERICVERSION_req_major=`expr $_AX_BOOST_BASE_TONUMERICVERSION_req : '\([[0-9]]*\)'`
+  AS_IF([test "x$_AX_BOOST_BASE_TONUMERICVERSION_req_major" = "x"],
+        [AC_MSG_ERROR([You should at least specify libboost major version])])
+  _AX_BOOST_BASE_TONUMERICVERSION_req_minor=`expr $_AX_BOOST_BASE_TONUMERICVERSION_req : '[[0-9]]*\.\([[0-9]]*\)'`
+  AS_IF([test "x$_AX_BOOST_BASE_TONUMERICVERSION_req_minor" = "x"],
+        [_AX_BOOST_BASE_TONUMERICVERSION_req_minor="0"])
+  _AX_BOOST_BASE_TONUMERICVERSION_req_sub_minor=`expr $_AX_BOOST_BASE_TONUMERICVERSION_req : '[[0-9]]*\.[[0-9]]*\.\([[0-9]]*\)'`
+  AS_IF([test "X$_AX_BOOST_BASE_TONUMERICVERSION_req_sub_minor" = "X"],
+        [_AX_BOOST_BASE_TONUMERICVERSION_req_sub_minor="0"])
+  _AX_BOOST_BASE_TONUMERICVERSION_RET=`expr $_AX_BOOST_BASE_TONUMERICVERSION_req_major \* 100000 \+  $_AX_BOOST_BASE_TONUMERICVERSION_req_minor \* 100 \+ $_AX_BOOST_BASE_TONUMERICVERSION_req_sub_minor`
+  AS_VAR_SET($1,$_AX_BOOST_BASE_TONUMERICVERSION_RET)
+])
+
+dnl Run the detection of boost should be run only if $want_boost
+AC_DEFUN([_AX_BOOST_BASE_RUNDETECT],[
+    _AX_BOOST_BASE_TONUMERICVERSION(WANT_BOOST_VERSION,[$1])
+    succeeded=no
+
+
+    AC_REQUIRE([AC_CANONICAL_HOST])
+    dnl On 64-bit systems check for system libraries in both lib64 and lib.
+    dnl The former is specified by FHS, but e.g. Debian does not adhere to
+    dnl this (as it rises problems for generic multi-arch support).
+    dnl The last entry in the list is chosen by default when no libraries
+    dnl are found, e.g. when only header-only libraries are installed!
+    AS_CASE([${host_cpu}],
+      [x86_64],[libsubdirs="lib64 libx32 lib lib64"],
+      [ppc64|s390x|sparc64|aarch64|ppc64le|riscv64],[libsubdirs="lib64 lib lib64"],
+      [libsubdirs="lib"]
+    )
+
+    dnl allow for real multi-arch paths e.g. /usr/lib/x86_64-linux-gnu. Give
+    dnl them priority over the other paths since, if libs are found there, they
+    dnl are almost assuredly the ones desired.
+    AS_CASE([${host_cpu}],
+      [i?86],[multiarch_libsubdir="lib/i386-${host_os}"],
+      [multiarch_libsubdir="lib/${host_cpu}-${host_os}"]
+    )
+
+    dnl first we check the system location for boost libraries
+    dnl this location ist chosen if boost libraries are installed with the --layout=system option
+    dnl or if you install boost with RPM
+    AS_IF([test "x$_AX_BOOST_BASE_boost_path" != "x"],[
+        AC_MSG_CHECKING([for boostlib >= $1 ($WANT_BOOST_VERSION) includes in "$_AX_BOOST_BASE_boost_path/include"])
+         AS_IF([test -d "$_AX_BOOST_BASE_boost_path/include" && test -r "$_AX_BOOST_BASE_boost_path/include"],[
+           AC_MSG_RESULT([yes])
+           BOOST_CPPFLAGS="-I$_AX_BOOST_BASE_boost_path/include"
+           for _AX_BOOST_BASE_boost_path_tmp in $multiarch_libsubdir $libsubdirs; do
+                AC_MSG_CHECKING([for boostlib >= $1 ($WANT_BOOST_VERSION) lib path in "$_AX_BOOST_BASE_boost_path/$_AX_BOOST_BASE_boost_path_tmp"])
+                AS_IF([test -d "$_AX_BOOST_BASE_boost_path/$_AX_BOOST_BASE_boost_path_tmp" && test -r "$_AX_BOOST_BASE_boost_path/$_AX_BOOST_BASE_boost_path_tmp" ],[
+                        AC_MSG_RESULT([yes])
+                        BOOST_LDFLAGS="-L$_AX_BOOST_BASE_boost_path/$_AX_BOOST_BASE_boost_path_tmp";
+                        break;
+                ],
+      [AC_MSG_RESULT([no])])
+           done],[
+      AC_MSG_RESULT([no])])
+    ],[
+        if test X"$cross_compiling" = Xyes; then
+            search_libsubdirs=$multiarch_libsubdir
+        else
+            search_libsubdirs="$multiarch_libsubdir $libsubdirs"
+        fi
+        for _AX_BOOST_BASE_boost_path_tmp in /usr /usr/local /opt /opt/local ; do
+            if test -d "$_AX_BOOST_BASE_boost_path_tmp/include/boost" && test -r "$_AX_BOOST_BASE_boost_path_tmp/include/boost" ; then
+                for libsubdir in $search_libsubdirs ; do
+                    if ls "$_AX_BOOST_BASE_boost_path_tmp/$libsubdir/libboost_"* >/dev/null 2>&1 ; then break; fi
+                done
+                BOOST_LDFLAGS="-L$_AX_BOOST_BASE_boost_path_tmp/$libsubdir"
+                BOOST_CPPFLAGS="-I$_AX_BOOST_BASE_boost_path_tmp/include"
+                break;
+            fi
+        done
+    ])
+
+    dnl overwrite ld flags if we have required special directory with
+    dnl --with-boost-libdir parameter
+    AS_IF([test "x$_AX_BOOST_BASE_boost_lib_path" != "x"],
+          [BOOST_LDFLAGS="-L$_AX_BOOST_BASE_boost_lib_path"])
+
+    AC_MSG_CHECKING([for boostlib >= $1 ($WANT_BOOST_VERSION)])
+    CPPFLAGS_SAVED="$CPPFLAGS"
+    CPPFLAGS="$CPPFLAGS $BOOST_CPPFLAGS"
+    export CPPFLAGS
+
+    LDFLAGS_SAVED="$LDFLAGS"
+    LDFLAGS="$LDFLAGS $BOOST_LDFLAGS"
+    export LDFLAGS
+
+    AC_REQUIRE([AC_PROG_CXX])
+    AC_LANG_PUSH(C++)
+        AC_COMPILE_IFELSE([_AX_BOOST_BASE_PROGRAM($WANT_BOOST_VERSION)],[
+        AC_MSG_RESULT(yes)
+    succeeded=yes
+    found_system=yes
+        ],[
+        ])
+    AC_LANG_POP([C++])
+
+
+
+    dnl if we found no boost with system layout we search for boost libraries
+    dnl built and installed without the --layout=system option or for a staged(not installed) version
+    if test "x$succeeded" != "xyes" ; then
+        CPPFLAGS="$CPPFLAGS_SAVED"
+        LDFLAGS="$LDFLAGS_SAVED"
+        BOOST_CPPFLAGS=
+        if test -z "$_AX_BOOST_BASE_boost_lib_path" ; then
+            BOOST_LDFLAGS=
+        fi
+        _version=0
+        if test -n "$_AX_BOOST_BASE_boost_path" ; then
+            if test -d "$_AX_BOOST_BASE_boost_path" && test -r "$_AX_BOOST_BASE_boost_path"; then
+                for i in `ls -d $_AX_BOOST_BASE_boost_path/include/boost-* 2>/dev/null`; do
+                    _version_tmp=`echo $i | sed "s#$_AX_BOOST_BASE_boost_path##" | sed 's/\/include\/boost-//' | sed 's/_/./'`
+                    V_CHECK=`expr $_version_tmp \> $_version`
+                    if test "x$V_CHECK" = "x1" ; then
+                        _version=$_version_tmp
+                    fi
+                    VERSION_UNDERSCORE=`echo $_version | sed 's/\./_/'`
+                    BOOST_CPPFLAGS="-I$_AX_BOOST_BASE_boost_path/include/boost-$VERSION_UNDERSCORE"
+                done
+                dnl if nothing found search for layout used in Windows distributions
+                if test -z "$BOOST_CPPFLAGS"; then
+                    if test -d "$_AX_BOOST_BASE_boost_path/boost" && test -r "$_AX_BOOST_BASE_boost_path/boost"; then
+                        BOOST_CPPFLAGS="-I$_AX_BOOST_BASE_boost_path"
+                    fi
+                fi
+                dnl if we found something and BOOST_LDFLAGS was unset before
+                dnl (because "$_AX_BOOST_BASE_boost_lib_path" = ""), set it here.
+                if test -n "$BOOST_CPPFLAGS" && test -z "$BOOST_LDFLAGS"; then
+                    for libsubdir in $libsubdirs ; do
+                        if ls "$_AX_BOOST_BASE_boost_path/$libsubdir/libboost_"* >/dev/null 2>&1 ; then break; fi
+                    done
+                    BOOST_LDFLAGS="-L$_AX_BOOST_BASE_boost_path/$libsubdir"
+                fi
+            fi
+        else
+            if test "x$cross_compiling" != "xyes" ; then
+                for _AX_BOOST_BASE_boost_path in /usr /usr/local /opt /opt/local ; do
+                    if test -d "$_AX_BOOST_BASE_boost_path" && test -r "$_AX_BOOST_BASE_boost_path" ; then
+                        for i in `ls -d $_AX_BOOST_BASE_boost_path/include/boost-* 2>/dev/null`; do
+                            _version_tmp=`echo $i | sed "s#$_AX_BOOST_BASE_boost_path##" | sed 's/\/include\/boost-//' | sed 's/_/./'`
+                            V_CHECK=`expr $_version_tmp \> $_version`
+                            if test "x$V_CHECK" = "x1" ; then
+                                _version=$_version_tmp
+                                best_path=$_AX_BOOST_BASE_boost_path
+                            fi
+                        done
+                    fi
+                done
+
+                VERSION_UNDERSCORE=`echo $_version | sed 's/\./_/'`
+                BOOST_CPPFLAGS="-I$best_path/include/boost-$VERSION_UNDERSCORE"
+                if test -z "$_AX_BOOST_BASE_boost_lib_path" ; then
+                    for libsubdir in $libsubdirs ; do
+                        if ls "$best_path/$libsubdir/libboost_"* >/dev/null 2>&1 ; then break; fi
+                    done
+                    BOOST_LDFLAGS="-L$best_path/$libsubdir"
+                fi
+            fi
+
+            if test -n "$BOOST_ROOT" ; then
+                for libsubdir in $libsubdirs ; do
+                    if ls "$BOOST_ROOT/stage/$libsubdir/libboost_"* >/dev/null 2>&1 ; then break; fi
+                done
+                if test -d "$BOOST_ROOT" && test -r "$BOOST_ROOT" && test -d "$BOOST_ROOT/stage/$libsubdir" && test -r "$BOOST_ROOT/stage/$libsubdir"; then
+                    version_dir=`expr //$BOOST_ROOT : '.*/\(.*\)'`
+                    stage_version=`echo $version_dir | sed 's/boost_//' | sed 's/_/./g'`
+                        stage_version_shorten=`expr $stage_version : '\([[0-9]]*\.[[0-9]]*\)'`
+                    V_CHECK=`expr $stage_version_shorten \>\= $_version`
+                    if test "x$V_CHECK" = "x1" && test -z "$_AX_BOOST_BASE_boost_lib_path" ; then
+                        AC_MSG_NOTICE(We will use a staged boost library from $BOOST_ROOT)
+                        BOOST_CPPFLAGS="-I$BOOST_ROOT"
+                        BOOST_LDFLAGS="-L$BOOST_ROOT/stage/$libsubdir"
+                    fi
+                fi
+            fi
+        fi
+
+        CPPFLAGS="$CPPFLAGS $BOOST_CPPFLAGS"
+        export CPPFLAGS
+        LDFLAGS="$LDFLAGS $BOOST_LDFLAGS"
+        export LDFLAGS
+
+        AC_LANG_PUSH(C++)
+            AC_COMPILE_IFELSE([_AX_BOOST_BASE_PROGRAM($WANT_BOOST_VERSION)],[
+            AC_MSG_RESULT(yes)
+        succeeded=yes
+        found_system=yes
+            ],[
+            ])
+        AC_LANG_POP([C++])
+    fi
+
+    if test "x$succeeded" != "xyes" ; then
+        if test "x$_version" = "x0" ; then
+            AC_MSG_NOTICE([[We could not detect the boost libraries (version $1 or higher). If you have a staged boost library (still not installed) please specify \$BOOST_ROOT in your environment and do not give a PATH to --with-boost option.  If you are sure you have boost installed, then check your version number looking in <boost/version.hpp>. See http://randspringer.de/boost for more documentation.]])
+        else
+            AC_MSG_NOTICE([Your boost libraries seems to old (version $_version).])
+        fi
+        # execute ACTION-IF-NOT-FOUND (if present):
+        ifelse([$3], , :, [$3])
+    else
+        AC_DEFINE(HAVE_BOOST,,[define if the Boost library is available])
+        # execute ACTION-IF-FOUND (if present):
+        ifelse([$2], , :, [$2])
+    fi
+
+    CPPFLAGS="$CPPFLAGS_SAVED"
+    LDFLAGS="$LDFLAGS_SAVED"
+
+])

--- a/src/bin/libint/comp_11_r1dotr1g12_11.h
+++ b/src/bin/libint/comp_11_r1dotr1g12_11.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <dgvertex.h>
 #include <rr.h>
 #include <rr.templ.h>

--- a/src/bin/libint/comp_11_r1dotr2g12_11.h
+++ b/src/bin/libint/comp_11_r1dotr2g12_11.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <dgvertex.h>
 #include <rr.h>
 #include <rr.templ.h>

--- a/src/bin/libint/comp_11_r2dotr2g12_11.h
+++ b/src/bin/libint/comp_11_r2dotr2g12_11.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <dgvertex.h>
 #include <rr.h>
 #include <rr.templ.h>

--- a/src/bin/libint/context.cc
+++ b/src/bin/libint/context.cc
@@ -18,6 +18,7 @@
  *
  */
 
+#include <cassert>
 #include <cstdio>
 #include <context.h>
 #include <codeblock.h>

--- a/src/bin/libint/dg.h
+++ b/src/bin/libint/dg.h
@@ -29,7 +29,7 @@
 #include <deque>
 #include <algorithm>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 
 #include <global_macros.h>
 #include <exception.h>

--- a/src/bin/libint/gauss.cc
+++ b/src/bin/libint/gauss.cc
@@ -18,6 +18,7 @@
  *
  */
 
+#include <cassert>
 #include <iostream>
 #include <sstream>
 #include <ctype.h>

--- a/src/bin/libint/generic_rr.h
+++ b/src/bin/libint/generic_rr.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <boost/type_traits/is_same.hpp>
 
 #include <dgvertex.h>

--- a/src/bin/libint/hrr.h
+++ b/src/bin/libint/hrr.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <rr.h>
 #include <integral.h>
 #include <algebra.h>

--- a/src/bin/libint/integral.h
+++ b/src/bin/libint/integral.h
@@ -21,6 +21,7 @@
 #ifndef _libint2_src_bin_libint_integral_h_
 #define _libint2_src_bin_libint_integral_h_
 
+#include <cassert>
 #include <smart_ptr.h>
 #include <dgvertex.h>
 #include <oper.h>

--- a/src/bin/libint/intset_to_ints.h
+++ b/src/bin/libint/intset_to_ints.h
@@ -22,7 +22,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <rr.h>
 #include <integral.h>
 #include <iter.h>

--- a/src/bin/libint/itr_11_twoprep_11.h
+++ b/src/bin/libint/itr_11_twoprep_11.h
@@ -26,7 +26,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <dgvertex.h>
 #include <rr.h>
 #include <twoprep_11_11.h>

--- a/src/bin/libint/quanta.h
+++ b/src/bin/libint/quanta.h
@@ -21,6 +21,7 @@
 #ifndef _libint2_src_bin_libint_quanta_h_
 #define _libint2_src_bin_libint_quanta_h_
 
+#include <cassert>
 #include <vector>
 #include <smart_ptr.h>
 #include <global_macros.h>

--- a/src/bin/libint/strategy.cc
+++ b/src/bin/libint/strategy.cc
@@ -66,7 +66,7 @@ namespace libint2 {
   template <class T> struct MasterStrategy;
 #if LIBINT_SHELLQUARTET_STRATEGY == LIBINT_SHELLQUARTET_STRATEGY_A0C0
   template <> struct MasterStrategy<TwoPRep_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_TwoPRep_11_sh,
     HRR_cd_11_TwoPRep_11_sh,
     Deriv_a_11_TwoPRep_11_sh,
@@ -82,7 +82,7 @@ namespace libint2 {
     > value;
   };
   template <> struct MasterStrategy<TwoPRep_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_TwoPRep_11_int,
     HRR_cd_11_TwoPRep_11_int,
     Deriv_a_11_TwoPRep_11_int,
@@ -99,7 +99,7 @@ namespace libint2 {
   };
 #else  // 0B0D strategy
   template <> struct MasterStrategy<TwoPRep_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_TwoPRep_11_sh,
     HRR_dc_11_TwoPRep_11_sh,
     Deriv_a_11_TwoPRep_11_sh,
@@ -115,7 +115,7 @@ namespace libint2 {
     > value;
   };
   template <> struct MasterStrategy<TwoPRep_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_TwoPRep_11_int,
     HRR_dc_11_TwoPRep_11_int,
     Deriv_a_11_TwoPRep_11_int,
@@ -134,7 +134,7 @@ namespace libint2 {
 
 #if LIBINT_SHELLQUARTET_STRATEGY == LIBINT_SHELLQUARTET_STRATEGY_A0C0
   template <> struct MasterStrategy<R12kG12_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_R12kG12_11_sh,
     HRR_cd_11_R12kG12_11_sh,
     VRR_a_11_R12kG12_11_sh,
@@ -142,7 +142,7 @@ namespace libint2 {
     > value;
   };
   template <> struct MasterStrategy<R12kG12_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_R12kG12_11_int,
     HRR_cd_11_R12kG12_11_int,
     VRR_a_11_R12kG12_11_int,
@@ -151,7 +151,7 @@ namespace libint2 {
   };
 #else // 0B0D strategy
   template <> struct MasterStrategy<R12kG12_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_R12kG12_11_sh,
     HRR_dc_11_R12kG12_11_sh,
     VRR_b_11_R12kG12_11_sh,
@@ -159,7 +159,7 @@ namespace libint2 {
     > value;
   };
   template <> struct MasterStrategy<R12kG12_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_R12kG12_11_int,
     HRR_dc_11_R12kG12_11_int,
     VRR_b_11_R12kG12_11_int,
@@ -169,56 +169,56 @@ namespace libint2 {
 #endif
 
   template <> struct MasterStrategy<R12kR12lG12_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_R12kR12lG12_11_sh,
     HRR_cd_11_R12kR12lG12_11_sh,
     CR_11_R12kR12lG12_11_sh
     > value;
   };
   template <> struct MasterStrategy<R12kR12lG12_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_R12kR12lG12_11_int,
     HRR_cd_11_R12kR12lG12_11_int,
     CR_11_R12kR12lG12_11_int
     > value;
   };
   template <> struct MasterStrategy<TiG12_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_TiG12_11_sh,
     HRR_cd_11_TiG12_11_sh,
     CR_11_TiG12_11_sh
     > value;
   };
   template <> struct MasterStrategy<TiG12_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_TiG12_11_int,
     HRR_cd_11_TiG12_11_int,
     CR_11_TiG12_11_int
     > value;
   };
   template <> struct MasterStrategy<G12TiG12_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_G12TiG12_11_sh,
     HRR_cd_11_G12TiG12_11_sh,
     CR_11_G12TiG12_11_sh
     > value;
   };
   template <> struct MasterStrategy<G12TiG12_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_G12TiG12_11_int,
     HRR_cd_11_G12TiG12_11_int,
     CR_11_G12TiG12_11_int
     > value;
   };
   template <> struct MasterStrategy<DivG12prime_xTx_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_DivG12prime_xTx_sh,
     HRR_cd_11_DivG12prime_xTx_sh,
     CR_11_DivG12prime_xTx_11_sh
     > value;
   };
   template <> struct MasterStrategy<DivG12prime_xTx_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_DivG12prime_xTx_int,
     HRR_cd_11_DivG12prime_xTx_int,
     CR_11_DivG12prime_xTx_11_int
@@ -226,26 +226,26 @@ namespace libint2 {
   };
 #if LIBINT_SHELLQUARTET_STRATEGY == LIBINT_SHELLQUARTET_STRATEGY_A0C0
   template <> struct MasterStrategy<DummySymmIntegral_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_Dummy_11_sh,
     HRR_cd_11_Dummy_11_sh
     > value;
   };
   template <> struct MasterStrategy<DummySymmIntegral_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ab_11_Dummy_11_int,
     HRR_cd_11_Dummy_11_int
     > value;
   };
 #else // 0B0D strategy
   template <> struct MasterStrategy<DummySymmIntegral_11_11_sq> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_Dummy_11_sh,
     HRR_dc_11_Dummy_11_sh
     > value;
   };
   template <> struct MasterStrategy<DummySymmIntegral_11_11_int> {
-    typedef mpl::list<
+    typedef boost::mpl::list<
     HRR_ba_11_Dummy_11_int,
     HRR_dc_11_Dummy_11_int
     > value;
@@ -254,41 +254,41 @@ namespace libint2 {
 
 #if LIBINT_SUPPORT_ONEBODYINTS
   template <> struct MasterStrategy<Overlap_1_1_sh_x> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGShell1d<CartesianAxis_X>,OverlapOper>
       > value;
     };
   template <> struct MasterStrategy<Overlap_1_1_sh_y> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGShell1d<CartesianAxis_Y>,OverlapOper>
       > value;
     };
   template <> struct MasterStrategy<Overlap_1_1_sh_z> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGShell1d<CartesianAxis_Z>,OverlapOper>
       > value;
     };
   template <> struct MasterStrategy<Overlap_1_1_int_x> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
       VRR_a_1_Overlap_1_int_x,
       VRR_b_1_Overlap_1_int_x
       > value;
     };
   template <> struct MasterStrategy<Overlap_1_1_int_y> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
       VRR_a_1_Overlap_1_int_y,
       VRR_b_1_Overlap_1_int_y
       > value;
     };
   template <> struct MasterStrategy<Overlap_1_1_int_z> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
       VRR_a_1_Overlap_1_int_z,
       VRR_b_1_Overlap_1_int_z
       > value;
     };
   // TODO create TwoCenterOSTactic to be able to optimally apply bidirectional strategies
   template <> struct MasterStrategy<ElecPot_1_1_sh> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
       HRR_ab_1_ElecPot_1_sh,
       HRR_ba_1_ElecPot_1_sh,
       CR_DerivGauss<ElecPot_1_1_sh,0,InBra>,
@@ -298,7 +298,7 @@ namespace libint2 {
       > value;
     };
   template <> struct MasterStrategy<ElecPot_1_1_int> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
       HRR_ab_1_ElecPot_1_int,
       HRR_ba_1_ElecPot_1_int,
       CR_DerivGauss<ElecPot_1_1_int,0,InBra>,
@@ -308,61 +308,61 @@ namespace libint2 {
       > value;
     };
   template <> struct MasterStrategy<Kinetic_1_1_sh> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGShell,KineticOper>
       > value;
     };
   template <> struct MasterStrategy<Kinetic_1_1_int> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_DerivGauss<Kinetic_1_1_int,0,InBra,trinvskip1_part,trinvskip1_where>,
         CR_DerivGauss<Kinetic_1_1_int,0,InKet,trinvskip1_part,trinvskip1_where>,
         CR_XYZ_1_1<CGF,KineticOper>
       > value;
     };
   template <> struct MasterStrategy<Kinetic_1_1_int_x> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_X>,KineticOper>
       > value;
     };
   template <> struct MasterStrategy<Kinetic_1_1_int_y> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_Y>,KineticOper>
       > value;
     };
   template <> struct MasterStrategy<Kinetic_1_1_int_z> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_Z>,KineticOper>
       > value;
     };
   template <> struct MasterStrategy<CMultipole_1_1_sh> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGShell,CartesianMultipoleOper<3u>>
       > value;
     };
   template <> struct MasterStrategy<CMultipole_1_1_int> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_DerivGauss<CMultipole_1_1_int,0,InBra>,
         CR_DerivGauss<CMultipole_1_1_int,0,InKet>,
         CR_XYZ_1_1<CGF,CartesianMultipoleOper<3u>>
       > value;
     };
   template <> struct MasterStrategy<CMultipole_1_1_int_x> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_X>,CartesianMultipoleOper<1u>>
       > value;
     };
   template <> struct MasterStrategy<CMultipole_1_1_int_y> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_Y>,CartesianMultipoleOper<1u>>
       > value;
     };
   template <> struct MasterStrategy<CMultipole_1_1_int_z> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         CR_XYZ_1_1<CGF1d<CartesianAxis_Z>,CartesianMultipoleOper<1u>>
       > value;
     };
   template <> struct MasterStrategy<SMultipole_1_1_sh> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
         HRR_ab_1_SMultipole_1_sh,
         HRR_ba_1_SMultipole_1_sh,
         CR_DerivGauss<SMultipole_1_1_sh,0,InBra>,
@@ -372,7 +372,7 @@ namespace libint2 {
       > value;
     };
   template <> struct MasterStrategy<SMultipole_1_1_int> {
-      typedef mpl::list<
+      typedef boost::mpl::list<
           HRR_ab_1_SMultipole_1_int,
           HRR_ba_1_SMultipole_1_int,
           CR_DerivGauss<SMultipole_1_1_int,0,InBra>,
@@ -384,7 +384,7 @@ namespace libint2 {
 #endif // LIBINT_SUPPORT_ONEBODYINTS
 
   template <typename T> struct MasterStrategy {
-      typedef mpl::list<> value;
+      typedef boost::mpl::list<> value;
   };
 
   /// transform<RRType> encapsulates RRType and the action associated with RRType

--- a/src/bin/libint/tactic.cc
+++ b/src/bin/libint/tactic.cc
@@ -18,6 +18,7 @@
  *
  */
 
+#include <cassert>
 #include <cstdlib>
 #include <ctime>
 #include <dg.h>

--- a/src/bin/libint/uncontract.h
+++ b/src/bin/libint/uncontract.h
@@ -25,7 +25,7 @@
 #include <string>
 #include <vector>
 #include <stdexcept>
-#include <assert.h>
+#include <cassert>
 #include <rr.h>
 #include <integral.h>
 #include <iter.h>

--- a/src/bin/libint/vrr_1_onep_1.h
+++ b/src/bin/libint/vrr_1_onep_1.h
@@ -21,6 +21,7 @@
 #ifndef _libint2_src_bin_libint_vrr1onep1_h_
 #define _libint2_src_bin_libint_vrr1onep1_h_
 
+#include <cassert>
 #include <generic_rr.h>
 #include <onep_1_1.h>
 

